src/clr-angular/utils/drag-and-drop/droppable/droppable.ts

```diff
@@ -0,0 +1,205 @@
+                return draggableGroup.indexOf(this._group) > -1;
+            } else {
+                return (this._group as string[]).filter(groupKey => draggableGroup.indexOf(groupKey) > -1).length > 0;
+            }
+        }
+    }
+
+    private isInDropArea(point: {pageX: number; pageY: number}): boolean {
+        if (!point) {
+            return false;
+        }
+
+        if (!this.clientRect) {
+            this.clientRect = this.domAdapter.clientRect(this.droppableEl);
+        }
+
+        if (point.pageX >= this.clientRect.left - this._dropTolerance.left &&
+            point.pageX <= this.clientRect.right + this._dropTolerance.right &&
+            point.pageY >= this.clientRect.top - this._dropTolerance.top &&
+            point.pageY <= this.clientRect.bottom + this._dropTolerance.bottom) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private onDragStart(dragStartEvent: ClrDragEvent<T>): void {
+        // Check draggable and droppable have a matching group key.
+        this.isDraggableMatch = this.checkGroupMatch(dragStartEvent.group);
+
+        // Subscribe to dragMoved and dragEnded only if draggable and droppable have a matching group key.
+        if (this.isDraggableMatch) {
+            this.dragStartEmitter.emit(dragStartEvent);
+            this.dragMoveSubscription = this.eventBus.dragMoved.subscribe((dragMoveEvent: ClrDragEvent<T>) => {
+                this.onDragMove(dragMoveEvent);
+            });
+            this.dragEndSubscription = this.eventBus.dragEnded.subscribe((dragEndEvent: ClrDragEvent<T>) => {
+                this.onDragEnd(dragEndEvent);
+            });
+        }
+    }
+
+    private onDragMove(dragMoveEvent: ClrDragEvent<T>): void {
+        if (!this._isDraggableOver && this.isInDropArea(dragMoveEvent.dropPointPosition)) {
+            this.isDraggableOver = true;
+            const dragEnterEvent = this.transformAndBroadcastEvent(dragMoveEvent, ClrDragEventType.DRAG_ENTER);
+            this.dragEnterEmitter.emit(dragEnterEvent);
+        } else if (this._isDraggableOver && !this.isInDropArea(dragMoveEvent.dropPointPosition)) {
+            this.isDraggableOver = false;
+            const dragLeaveEvent = this.transformAndBroadcastEvent(dragMoveEvent, ClrDragEventType.DRAG_LEAVE);
+            this.dragLeaveEmitter.emit(dragLeaveEvent);
+        }
+
+        this.dragMoveEmitter.emit(dragMoveEvent);
+    }
+
+    private onDragEnd(dragEndEvent: ClrDragEvent<T>): void {
+        if (this._isDraggableOver) {
+            if (dragEndEvent.ghostElement) {
+                // By this point draggable ghost component would have been destroyed,
+                // but the element would be active until its animation completes.
+                // As such, once the ghost is dropped over, we will give it "dropped" class.
+                this.renderer.addClass(dragEndEvent.ghostElement, "dropped");
```

Comment:
I believe this should be in the ghost component, not here. You are broadcasting a DROP event, so you can have the ghost just listen for it and cancel the animation/disappear when it receives the event. You wouldn't have to pass around the native element for the ghost anymore.

Reply:
This one cannot be done in the ghost component because any subscription to the `drop` event (or `dragEnd`) is ineffective or invalid as the component is already destroyed at this point. The ghost component can only subscribe to the `dragMove` event. I've annotated why it's here though. But also, I am open to other suggestions or improvements.

Reply:
My first intuition would be to just put that code in the ghost component's `ngOnDestroy()`, then. But I realize this wouldn't work since you only want the class to be applied if it was successfully dropped. This is fine for now. 👍